timeseries/src/main/scala/com/criteo/cuttle/timeseries/TimeSeriesScheduler.scala

```diff
@@ -287,66 +287,91 @@ case class TimeSeriesScheduler() extends Scheduler[TimeSeries] with TimeSeriesAp
       }
     }
 
-    def go(running: Set[Run]): Unit = {
-      val (completed, stillRunning) = running.partition(_._3.isCompleted)
+    def mainLoop(running: Set[Run]): Unit = {
+      val (completed, stillRunning) = running.partition {
+        case(_,_,effect) => effect.isCompleted
+      }
+
       val now = Instant.now
       val (stateSnapshot, completedBackfills, toRun) = atomic { implicit txn =>
-        completed.foreach {
-          case (job, context, future) =>
-            val jobState = if (future.value.get.isSuccess) Done else Todo(context.backfill)
-            _state() = _state() + (job ->
-              (_state().apply(job).update(context.toInterval, jobState)))
-        }
+        val (stateSnapshot, newBackfills, completedBackfills) =
+          computeNextState(workflow, now, _state(), _backfills(), completed)
 
-        val oldBackfills = _backfills.snapshot
-        _backfills.retain { bf =>
-          val itvl = Interval(bf.start, bf.end)
-          !bf.jobs.forall(job => _state().apply(job).intersect(itvl).toList.forall(_._2 == Done))
-        }
-        val _toRun = next(workflow, _state(), now)
+        val toRun = jobsToRun(workflow, stateSnapshot, now)
 
-        (_state(), oldBackfills -- _backfills.snapshot, _toRun)
+        _state() = stateSnapshot
+        _backfills() = newBackfills
+
+        (stateSnapshot, completedBackfills, toRun)
       }
 
       val newExecutions = executor.runAll(toRun)
 
       atomic { implicit txn =>
-        _state() = newExecutions.foldLeft(_state()) { (st, x) =>
-          val (execution, _) = x
-          st + (execution.job ->
-            st(execution.job).update(execution.context.toInterval, Running(execution.id)))
+        _state() = newExecutions.foldLeft(_state()) {
+          case (st, (execution, _)) =>
```

Comment:
Apparently this works.
 http://scala-lang.org/files/archive/spec/2.11/08-pattern-matching.html#pattern-matching-anonymous-functions
I did not know that.